libraries/AP_RangeFinder/RangeFinder.h

```diff
@@ -35,7 +35,7 @@ class AP_RangeFinder_Backend;
 class RangeFinder
 {
     friend class AP_RangeFinder_Backend;
-
+    friend class AP_RangeFinder_UAVCAN;
```

Comment:
Why do you need a friend class? Just can't see what you need/that you aren't getting via the backend?

Reply:
I need it to get list of drivers, I scan through them and initialise in the backend itself.

Reply:
I will place a comment for reference.

Reply:
But why not allocated them in the loop in RangeFinder.cpp like normal ones, then you can simply leave them with an unassigned flag? This resolves a couple of things for you:

- `num_instances` has been dropped from your patches, which means it won't be reporting correctly for UAVCAN that I can see (which is a real bug)
- Your fix of adding a backend type flag works, but if you allocate the drivers here, you don't need to introduce the extra member everywhere, and only pay the cost of the bool flag on the UAVCAN instances
- It would allow you to remove the friend class, and use the param instance you are already passing to the backend

Reply:
Yes, you are right. Except I will still need to be friends with frontend to get the list of drivers unless I maintain a list in the backend itself.

Reply:
Yeah that's true. Each driver would have to register itself onto a static list, which is going to be messy as well... (Would get you out of friends again though :D)

Either way `num_instances` needs help :)

Reply:
Yes, num_instances was a real bug, thanks for pointing it out. Fixed in https://github.com/ArduPilot/ardupilot/pull/11017/commits/14c191bd4a8eea0007274009c8e92386515258fd